[Bug] range index 残留索引导致统计不准确

## How to reproduce ( 复现步骤 )

当前vertex写入相同id数据时，老索引并不会删除，会在后续查询是通过condtion条件过滤进行删除

但是range索引会存在问题，例如我们写入2000条数据

```java

        SchemaManager schema = graph.schema();

        schema.vertexLabel("sj").properties("name", "update_time")
              .primaryKeys("name").create();

        schema.indexLabel("sj_update_time_range").onV("sj").range()
              .by("update_time").create();

        for (int i=1; i<=2000; i++) {
            graph.addVertex(T.label, "sj", "name", "sj"+i,
                            "update_time", i);
        }
        graph.tx().commit();

        List<Vertex> vertices;
        long count = graph.traversal().V().has("sj", "update_time", ConditionP.gt(0))
                                          .count().next();
        Assert.assertEquals(2000L, count);
```

让后更新 `update_time`：

```java
        for (int i=1; i<=2000; i++) {
            graph.addVertex(T.label, "sj", "name", "sj"+i,
                            "update_time", 2*i);
        }
        graph.tx().commit();

        count = graph.traversal().V().has("sj", "update_time", ConditionP.gt(0))
                                     .count().next();
        Assert.assertEquals(2000L, count);
```

查询结果是3500：

```java
Exception in thread "main" java.lang.AssertionError: expected:<2000> but was:<3500>
 at org.junit.Assert.fail(Assert.java:88)
```

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java
@@ -170,7 +170,7 @@ public final class ConditionQuery extends IdQuery {
         this.element2IndexValueMap.removeElementLeftIndex(elementId);
     }
 
-    public boolean elementHasLeftIndex(Id elementId) {
+    public boolean existLeftIndex(Id elementId) {
         return this.getElementLeftIndex(elementId) != null;
     }
 

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java
@@ -19,6 +19,7 @@
 
 package com.baidu.hugegraph.backend.query;
 
+import java.math.BigDecimal;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Date;
@@ -37,6 +38,7 @@ import com.baidu.hugegraph.backend.query.Condition.Relation;
 import com.baidu.hugegraph.backend.query.Condition.RelationType;
 import com.baidu.hugegraph.perf.PerfUtil.Watched;
 import com.baidu.hugegraph.structure.HugeElement;
+import com.baidu.hugegraph.structure.HugeProperty;
 import com.baidu.hugegraph.type.HugeType;
 import com.baidu.hugegraph.type.define.CollectionType;
 import com.baidu.hugegraph.type.define.HugeKeys;
@@ -46,6 +48,7 @@ import com.baidu.hugegraph.util.NumericUtil;
 import com.baidu.hugegraph.util.collection.CollectionFactory;
 import com.google.common.base.Function;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
 
 public final class ConditionQuery extends IdQuery {
 
@@ -56,6 +59,7 @@ public final class ConditionQuery extends IdQuery {
 
     private OptimizedType optimizedType = OptimizedType.NONE;
     private Function<HugeElement, Boolean> resultsFilter = null;
+    private Element2IndexValueMap element2IndexValueMap = null;
 
     public ConditionQuery(HugeType resultType) {
         super(resultType);
@@ -140,6 +144,34 @@ public final class ConditionQuery extends IdQuery {
 
     public void resetConditions() {
         this.conditions = new LinkedHashSet<>();
+    }
+
+    public void recordIndexValue(Id propertyId, Id id, Object indexValue) {
+        this.ensureElement2IndexValueMap();
+        this.element2IndexValueMap.addIndexValue(propertyId, id, indexValue);
+    }
+
+    public void selectedIndexField(Id indexField) {
+        this.ensureElement2IndexValueMap();
+        this.element2IndexValueMap.selectedIndexField(indexField);
+    }
+
+    public Set<LeftIndex> getElementLeftIndex(Id elementId) {
+        if (this.element2IndexValueMap == null) {
+            return null;
+        }
+        return this.element2IndexValueMap.getLeftIndex(elementId);
+    }
+
+    public void removeElementLeftIndex(Id elementId) {
+        if (this.element2IndexValueMap == null) {
+            return;
+        }
+        this.element2IndexValueMap.removeElementLeftIndex(elementId);
+    }
+
+    public boolean elementHasLeftIndex(Id elementId) {
+        return this.getElementLeftIndex(elementId) != null;
     }
 
     public List<Condition.Relation> relations() {
@@ -417,7 +449,6 @@ public final class ConditionQuery extends IdQuery {
         return false;
     }
 
-
     public boolean matchUserpropKeys(List<Id> keys) {
         Set<Id> conditionKeys = this.userpropKeys();
         return keys.size() > 0 && conditionKeys.containsAll(keys);
@@ -455,13 +486,13 @@ public final class ConditionQuery extends IdQuery {
         if (this.resultsFilter != null) {
             return this.resultsFilter.apply(element);
         }
-
+        boolean valid = true;
         for (Condition cond : this.conditions()) {
-            if (!cond.test(element)) {
-                return false;
-            }
+            valid &= cond.test(element);
+            valid &= (this.element2IndexValueMap == null ||
+                      this.element2IndexValueMap.validRangeIndex(element,  cond));
         }
-        return true;
+        return valid;
     }
 
     public void checkFlattened() {
@@ -530,6 +561,24 @@ public final class ConditionQuery extends IdQuery {
         }
     }
 
+    public ConditionQuery originConditionQuery() {
+        Query originQuery = this.originQuery();
+        if (!(originQuery instanceof ConditionQuery)) {
+            return null;
+        }
+
+        while (originQuery.originQuery() instanceof ConditionQuery) {
+            originQuery = originQuery.originQuery();
+        }
+        return (ConditionQuery) originQuery;
+    }
+
+    private void ensureElement2IndexValueMap() {
+        if (this.element2IndexValueMap == null) {
+            this.element2IndexValueMap = new Element2IndexValueMap();
+        }
+    }
+
     public static String concatValues(List<Object> values) {
         List<Object> newValues = new ArrayList<>(values.size());
         for (Object v : values) {
@@ -567,5 +616,150 @@ public final class ConditionQuery extends IdQuery {
         SORT_KEYS,
         INDEX,
         INDEX_FILTER
+    }
+
+    public static final class Element2IndexValueMap {
+
+        private final Map<Id, Set<LeftIndex>> leftIndexMap;
+        private final Map<Id, Map<Id, Set<Object>>> filed2IndexValues;
+        private Id selectedIndexField;
+
+        public Element2IndexValueMap() {
+            this.filed2IndexValues = new HashMap<>();
+            this.leftIndexMap = new HashMap<>();
+        }
+
+        public void addIndexValue(Id indexField, Id elementId,
+                                  Object indexValue) {
+            if (!this.filed2IndexValues.containsKey(indexField)) {
+                this.filed2IndexValues.put(indexField, new HashMap<>());
+            }
+            Map<Id, Set<Object>> element2IndexValueMap =
+                                 this.filed2IndexValues.get(indexField);
+            if (element2IndexValueMap.containsKey(elementId)) {
+                element2IndexValueMap.get(elementId).add(indexValue);
+            } else {
+                element2IndexValueMap.put(elementId,
+                                          Sets.newHashSet(indexValue));
+            }
+        }
+
+        public void selectedIndexField(Id indexField) {
+            this.selectedIndexField = indexField;
+        }
+
+        public Set<Object> removeIndexValues(Id indexField, Id elementId) {
+            if (!this.filed2IndexValues.containsKey(indexField)) {
+                return null;
+            }
+            return this.filed2IndexValues.get(indexField).get(elementId);
+        }
+
+        public void addLeftIndex(Id indexField, Set<Object> indexValues,
+                                 Id elementId) {
+            LeftIndex leftIndex = new LeftIndex(indexValues, indexField);
+            if (this.leftIndexMap.containsKey(elementId)) {
+                this.leftIndexMap.get(elementId).add(leftIndex);
+            } else {
+                this.leftIndexMap.put(elementId, Sets.newHashSet(leftIndex));
+            }
+        }
+
+        public Set<LeftIndex> getLeftIndex(Id elementId) {
+            return this.leftIndexMap.get(elementId);
+        }
+
+        public void removeElementLeftIndex(Id elementId) {
+            this.leftIndexMap.remove(elementId);
+        }
+
+        public boolean validRangeIndex(HugeElement element, Condition cond) {
+            // Not UserpropRelation
+            if (!(cond instanceof Condition.UserpropRelation)) {
+                return true;
+            }
+
+            Condition.UserpropRelation propRelation =
+                                       (Condition.UserpropRelation) cond;
+            Id propId = propRelation.key();
+            Set<Object> fieldValues = this.removeIndexValues(propId,
+                                                             element.id());
+            if (fieldValues == null) {
+                // Not range index
+                return true;
+            }
+
+            HugeProperty<Object> hugeProperty = element.getProperty(propId);
+            if (hugeProperty == null) {
+                // Property value has been deleted
+                this.addLeftIndex(propId, fieldValues, element.id());
+                return false;
+            }
+
+            /*
+             * NOTE: If success remove means has correct index,
+             * we should add left index values to left index map
+             * waiting to be removed
+             */
+            boolean hasRightValue = removeValue(fieldValues, hugeProperty.value());
+            if (fieldValues.size() > 0) {
+                this.addLeftIndex(propId, fieldValues, element.id());
+            }
+
+            /*
+             * NOTE: When query by more than one range index field,
+             * if current field is not the selected one, it can only be used to
+             * determine whether the index values matched, can't determine
+             * the element is valid or not
+             */
+            if (this.selectedIndexField != null) {
+                return !propId.equals(this.selectedIndexField) || hasRightValue;
+            }
+
+            return hasRightValue;
+        }
+
+        private static boolean removeValue(Set<Object> values, Object value){
+            for (Object compareValue : values) {
+                if (numberEquals(compareValue, value)) {
+                    values.remove(compareValue);
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private static boolean numberEquals(Object number1, Object number2) {
+            // Same class compare directly
+            if (number1.getClass().equals(number2.getClass())) {
+                return number1.equals(number2);
+            }
+
+            // Otherwise convert to BigDecimal to make two numbers comparable
+            Number n1 = NumericUtil.convertToNumber(number1);
+            Number n2 = NumericUtil.convertToNumber(number2);
+            BigDecimal b1 = new BigDecimal(n1.doubleValue());
+            BigDecimal b2 = new BigDecimal(n2.doubleValue());
+            return b1.compareTo(b2) == 0;
+        }
+    }
+
+    public static final class LeftIndex {
+
+        private final Set<Object> indexFieldValues;
+        private final Id indexField;
+
+        public LeftIndex(Set<Object> indexFieldValues, Id indexField) {
+            this.indexFieldValues = indexFieldValues;
+            this.indexField = indexField;
+        }
+
+        public Set<Object> indexFieldValues() {
+            return indexFieldValues;
+        }
+
+        public Id indexField() {
+            return indexField;
+        }
     }
 }

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java
@@ -120,6 +120,7 @@ public class GraphIndexTransaction extends AbstractTransaction {
 
     protected Id asyncRemoveIndexLeft(ConditionQuery query,
                                       HugeElement element) {
+        LOG.info("Remove left index: {}, query: {}", element, query);
         RemoveLeftIndexJob job = new RemoveLeftIndexJob(query, element);
         HugeTask<?> task = EphemeralJobBuilder.of(this.graph())
                                               .name(element.id().asString())
@@ -589,6 +590,7 @@ public class GraphIndexTransaction extends AbstractTransaction {
             IdHolder holder = this.doIndexQuery(indexLabel, query);
             if (resultHolder == null) {
                 resultHolder = holder;
+                this.storeSelectedIndexField(indexLabel, query);
             }
             assert this.indexIntersectThresh > 0; // default value is 1000
             Set<Id> ids = ((BatchIdHolder) holder).peekNext(
@@ -600,6 +602,7 @@ public class GraphIndexTransaction extends AbstractTransaction {
             } else if (filtering) {
                 assert ids.size() < this.indexIntersectThresh;
                 resultHolder = holder;
+                this.storeSelectedIndexField(indexLabel, query);
                 break;
             } else {
                 if (intersectIds == null) {
@@ -618,6 +621,20 @@ public class GraphIndexTransaction extends AbstractTransaction {
         } else {
             assert intersectIds != null;
             return new FixedIdHolder(queries.asJointQuery(), intersectIds);
+        }
+    }
+
+    private void storeSelectedIndexField(IndexLabel indexLabel,
+                                         ConditionQuery query) {
+        // Only store range index field
+        if (!indexLabel.indexType().isRange()) {
+            return;
+        }
+
+        ConditionQuery originConditionQuery =
+                       query.originConditionQuery();
+        if (originConditionQuery != null) {
+            originConditionQuery.selectedIndexField(indexLabel.indexField());
         }
     }
 
@@ -660,12 +677,26 @@ public class GraphIndexTransaction extends AbstractTransaction {
                     this.removeExpiredIndexIfNeeded(index, query.showExpired());
                     ids.addAll(index.elementIds());
                     Query.checkForceCapacity(ids.size());
+                    this.recordIndexValue(query, index);
                 }
                 return ids;
             } finally {
                 locks.unlock();
             }
         });
+    }
+
+    private void recordIndexValue(ConditionQuery query, HugeIndex index) {
+        if (!shouldRecordIndexValue(query, index)) {
+            return;
+        }
+
+        ConditionQuery originQuery = query.originConditionQuery();
+        Id fieldId = index.indexLabel().indexField();
+        for (Id id : index.elementIds()) {
+            Object value = index.indexLabel().validValue(index.fieldValues());
+            originQuery.recordIndexValue(fieldId, id, value);
+        }
     }
 
     @Watched(prefix = "index")
@@ -689,6 +720,7 @@ public class GraphIndexTransaction extends AbstractTransaction {
                     break;
                 }
                 Query.checkForceCapacity(ids.size());
+                this.recordIndexValue(query, index);
             }
             // If there is no data, the entries is not a Metadatable object
             if (ids.isEmpty()) {
@@ -1091,6 +1123,13 @@ public class GraphIndexTransaction extends AbstractTransaction {
             return true;
         }
         return false;
+    }
+
+    private static boolean shouldRecordIndexValue(ConditionQuery query,
+                                                  HugeIndex index) {
+        // Currently, only range index has problems
+        return query.originQuery() instanceof ConditionQuery &&
+               index.indexLabel().indexType().isRange();
     }
 
     private static IndexQueries constructJointSecondaryQueries(
@@ -1656,6 +1695,7 @@ public class GraphIndexTransaction extends AbstractTransaction {
         private final ConditionQuery query;
         private final HugeElement element;
         private GraphIndexTransaction tx;
+        private Set<ConditionQuery.LeftIndex> leftIndexes;
 
         private RemoveLeftIndexJob(ConditionQuery query, HugeElement element) {
             E.checkArgumentNotNull(query, "query");
@@ -1663,6 +1703,7 @@ public class GraphIndexTransaction extends AbstractTransaction {
             this.query = query;
             this.element = element;
             this.tx = null;
+            this.leftIndexes = query.getElementLeftIndex(element.id());
         }
 
         @Override
@@ -1716,53 +1757,44 @@ public class GraphIndexTransaction extends AbstractTransaction {
 
         private long processRangeIndexLeft(ConditionQuery query,
                                            HugeElement element) {
-            GraphIndexTransaction tx = this.tx;
-            AbstractSerializer serializer = tx.serializer;
             long count = 0;
-            // Construct index ConditionQuery
-            Set<MatchedIndex> matchedIndexes = tx.collectMatchedIndexes(query);
-            IndexQueries queries = null;
-            Id elementLabelId = element.schemaLabel().id();
-            for (MatchedIndex index : matchedIndexes) {
-                if (index.schemaLabel().id().equals(elementLabelId)) {
-                    queries = index.constructIndexQueries(query);
-                    break;
-                }
+            if (this.leftIndexes == null) {
+                return count;
             }
 
-            E.checkState(queries != null,
-                         "Can't construct left-index query for '%s'", query);
+            for (ConditionQuery.LeftIndex leftIndex : this.leftIndexes) {
+                Set<Object> indexValues = leftIndex.indexFieldValues();
+                IndexLabel indexLabel = this.findMatchedIndexLabel(query,
+                                                                   leftIndex);
+                assert indexLabel != null;
 
-            for (ConditionQuery q : queries.values()) {
-                if (!q.resultType().isRangeIndex()) {
-                    continue;
-                }
-                // Query and delete index equals element id
-                Iterator<BackendEntry> it = tx.query(q).iterator();
-                try {
-                    while (it.hasNext()) {
-                        HugeIndex index = serializer.readIndex(graph(), q,
-                                                               it.next());
-                        tx.removeExpiredIndexIfNeeded(index, q.showExpired());
-                        if (index.elementIds().contains(element.id())) {
-                            index.resetElementIds();
-                            index.elementIds(element.id());
-                            tx.doEliminate(serializer.writeIndex(index));
-                            tx.commit();
-                            // If deleted by error, re-add deleted index again
-                            if (this.deletedByError(query, element)) {
-                                tx.doAppend(serializer.writeIndex(index));
-                                tx.commit();
-                            } else {
-                                count++;
-                            }
-                        }
-                    }
-                } finally {
-                    CloseableIterator.closeIterator(it);
+                AbstractSerializer serializer = this.tx.serializer;
+                for (Object value : indexValues) {
+                    HugeIndex index = new HugeIndex(this.graph(), indexLabel);
+                    index.elementIds(element.id());
+                    index.fieldValues(value);
+                    this.tx.doEliminate(serializer.writeIndex(index));
+                    count++;
                 }
             }
+            // Remove LeftIndex after constructing remove job
+            this.query.removeElementLeftIndex(element.id());
+            this.tx.commit();
             return count;
+        }
+
+        private IndexLabel findMatchedIndexLabel(ConditionQuery query,
+                                                 ConditionQuery.LeftIndex
+                                                 leftIndex) {
+            Set<MatchedIndex> matchedIndexes = this.tx.collectMatchedIndexes(query);
+            for (MatchedIndex index : matchedIndexes) {
+               for (IndexLabel label : index.indexLabels()){
+                   if (label.indexField().equals(leftIndex.indexField())){
+                       return label;
+                   }
+               }
+            }
+            return null;
         }
 
         private long processSecondaryOrSearchIndexLeft(ConditionQuery query,

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java
@@ -321,7 +321,7 @@ public class GraphTransaction extends IndexableTransaction {
         }
 
         if (this.removeLeftIndexOnOverwrite) {
-            this.removeLeftIndex(addedVertices);
+            this.removeLeftIndexIfNeeded(addedVertices);
         }
 
         // Do vertex update
@@ -356,25 +356,6 @@ public class GraphTransaction extends IndexableTransaction {
             // Update index of edge
             this.indexTx.updateEdgeIndex(e, false);
             this.indexTx.updateLabelIndex(e, false);
-        }
-    }
-
-    private void removeLeftIndex(Map<Id, HugeVertex> vertices) {
-        Set<Id> ids = vertices.keySet();
-        if (ids.isEmpty()) {
-            return;
-        }
-        IdQuery idQuery = new IdQuery(HugeType.VERTEX, ids);
-        Iterator<HugeVertex> results = this.queryVerticesFromBackend(idQuery);
-        try {
-            while (results.hasNext()) {
-                HugeVertex existedVertex = results.next();
-                HugeVertex newVertex = vertices.get(existedVertex.id());
-                checkLabel(existedVertex, newVertex);
-                this.indexTx.updateVertexIndex(existedVertex, true);
-            }
-        } finally {
-            CloseableIterator.closeIterator(results);
         }
     }
 
@@ -1575,7 +1556,14 @@ public class GraphTransaction extends IndexableTransaction {
             }
             HugeVertex existedVertex = results.next();
             HugeVertex newVertex = vertices.get(existedVertex.id());
-            checkLabel(existedVertex, newVertex);
+            if (!existedVertex.label().equals(newVertex.label())) {
+                throw new HugeException(
+                          "The newly added vertex with id:'%s' label:'%s' " +
+                          "is not allowed to insert, because already exist " +
+                          "a vertex with same id and different label:'%s'",
+                          newVertex.id(), newVertex.label(),
+                          existedVertex.label());
+            }
         } finally {
             CloseableIterator.closeIterator(results);
         }
@@ -1617,6 +1605,23 @@ public class GraphTransaction extends IndexableTransaction {
         } catch (Throwable e) {
             this.locksTable.unlock();
             throw e;
+        }
+    }
+
+    private void removeLeftIndexIfNeeded(Map<Id, HugeVertex> vertices) {
+        Set<Id> ids = vertices.keySet();
+        if (ids.isEmpty()) {
+            return;
+        }
+        IdQuery idQuery = new IdQuery(HugeType.VERTEX, ids);
+        Iterator<HugeVertex> results = this.queryVerticesFromBackend(idQuery);
+        try {
+            while (results.hasNext()) {
+                HugeVertex existedVertex = results.next();
+                this.indexTx.updateVertexIndex(existedVertex, true);
+            }
+        } finally {
+            CloseableIterator.closeIterator(results);
         }
     }
 
@@ -1665,7 +1670,7 @@ public class GraphTransaction extends IndexableTransaction {
 
         ConditionQuery cq = (ConditionQuery) query;
         if (cq.optimized() == OptimizedType.NONE || cq.test(elem)) {
-            if (cq.elementHasLeftIndex(elem.id())) {
+            if (cq.existLeftIndex(elem.id())) {
                 /*
                  * Both have correct and left index, wo should return true
                  * but also needs to cleaned up left index
@@ -2040,18 +2045,6 @@ public class GraphTransaction extends IndexableTransaction {
                     CloseableIterator.closeIterator(iter);
                 }
             } while (page != null);
-        }
-    }
-
-    private static void checkLabel(HugeVertex existedVertex,
-                                   HugeVertex newVertex) {
-        if (!existedVertex.label().equals(newVertex.label())) {
-            throw new HugeException(
-                      "The newly added vertex with id:'%s' label:'%s' " +
-                      "is not allowed to insert, because already exist " +
-                      "a vertex with same id and different label:'%s'",
-                      newVertex.id(), newVertex.label(),
-                      existedVertex.label());
         }
     }
 }

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java
@@ -125,6 +125,7 @@ public class GraphTransaction extends IndexableTransaction {
     private final boolean checkCustomVertexExist;
     private final boolean checkAdjacentVertexExist;
     private final boolean lazyLoadAdjacentVertex;
+    private final boolean removeLeftIndexOnOverwrite;
     private final boolean ignoreInvalidEntry;
     private final int commitPartOfAdjacentEdges;
     private final int batchSize;
@@ -148,6 +149,8 @@ public class GraphTransaction extends IndexableTransaction {
              conf.get(CoreOptions.VERTEX_ADJACENT_VERTEX_EXIST);
         this.lazyLoadAdjacentVertex =
              conf.get(CoreOptions.VERTEX_ADJACENT_VERTEX_LAZY);
+        this.removeLeftIndexOnOverwrite =
+             conf.get(CoreOptions.VERTEX_REMOVE_LEFT_INDEX);
         this.commitPartOfAdjacentEdges =
              conf.get(CoreOptions.VERTEX_PART_EDGE_COMMIT_SIZE);
         this.ignoreInvalidEntry =
@@ -316,6 +319,11 @@ public class GraphTransaction extends IndexableTransaction {
         if (this.checkCustomVertexExist) {
             this.checkVertexExistIfCustomizedId(addedVertices);
         }
+
+        if (this.removeLeftIndexOnOverwrite) {
+            this.removeLeftIndex(addedVertices);
+        }
+
         // Do vertex update
         for (HugeVertex v : addedVertices.values()) {
             assert !v.removed();
@@ -348,6 +356,25 @@ public class GraphTransaction extends IndexableTransaction {
             // Update index of edge
             this.indexTx.updateEdgeIndex(e, false);
             this.indexTx.updateLabelIndex(e, false);
+        }
+    }
+
+    private void removeLeftIndex(Map<Id, HugeVertex> vertices) {
+        Set<Id> ids = vertices.keySet();
+        if (ids.isEmpty()) {
+            return;
+        }
+        IdQuery idQuery = new IdQuery(HugeType.VERTEX, ids);
+        Iterator<HugeVertex> results = this.queryVerticesFromBackend(idQuery);
+        try {
+            while (results.hasNext()) {
+                HugeVertex existedVertex = results.next();
+                HugeVertex newVertex = vertices.get(existedVertex.id());
+                checkLabel(existedVertex, newVertex);
+                this.indexTx.updateVertexIndex(existedVertex, true);
+            }
+        } finally {
+            CloseableIterator.closeIterator(results);
         }
     }
 
@@ -1548,14 +1575,7 @@ public class GraphTransaction extends IndexableTransaction {
             }
             HugeVertex existedVertex = results.next();
             HugeVertex newVertex = vertices.get(existedVertex.id());
-            if (!existedVertex.label().equals(newVertex.label())) {
-                throw new HugeException(
-                          "The newly added vertex with id:'%s' label:'%s' " +
-                          "is not allowed to insert, because already exist " +
-                          "a vertex with same id and different label:'%s'",
-                          newVertex.id(), newVertex.label(),
-                          existedVertex.label());
-            }
+            checkLabel(existedVertex, newVertex);
         } finally {
             CloseableIterator.closeIterator(results);
         }
@@ -1645,6 +1665,14 @@ public class GraphTransaction extends IndexableTransaction {
 
         ConditionQuery cq = (ConditionQuery) query;
         if (cq.optimized() == OptimizedType.NONE || cq.test(elem)) {
+            if (cq.elementHasLeftIndex(elem.id())) {
+                /*
+                 * Both have correct and left index, wo should return true
+                 * but also needs to cleaned up left index
+                 */
+                this.indexTx.asyncRemoveIndexLeft(cq, elem);
+            }
+
             /* Return true if:
              * 1.not query by index or by primary-key/sort-key
              *   (cq.optimized() == 0 means query just by sysprop)
@@ -1654,7 +1682,6 @@ public class GraphTransaction extends IndexableTransaction {
         }
 
         if (cq.optimized() == OptimizedType.INDEX) {
-            LOG.info("Remove left index: {}, query: {}", elem, cq);
             this.indexTx.asyncRemoveIndexLeft(cq, elem);
         }
         return false;
@@ -2013,6 +2040,18 @@ public class GraphTransaction extends IndexableTransaction {
                     CloseableIterator.closeIterator(iter);
                 }
             } while (page != null);
+        }
+    }
+
+    private static void checkLabel(HugeVertex existedVertex,
+                                   HugeVertex newVertex) {
+        if (!existedVertex.label().equals(newVertex.label())) {
+            throw new HugeException(
+                      "The newly added vertex with id:'%s' label:'%s' " +
+                      "is not allowed to insert, because already exist " +
+                      "a vertex with same id and different label:'%s'",
+                      newVertex.id(), newVertex.label(),
+                      existedVertex.label());
         }
     }
 }

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java
@@ -352,6 +352,14 @@ public class CoreOptions extends OptionHolder {
                     false
             );
 
+    public static final ConfigOption<Boolean> VERTEX_REMOVE_LEFT_INDEX =
+            new ConfigOption<>(
+                    "vertex.remove_left_index_at_overwrite",
+                    "Whether remove left index at overwrite.",
+                    disallowEmpty(),
+                    false
+            );
+
     public static final ConfigOption<Boolean> VERTEX_ADJACENT_VERTEX_EXIST =
             new ConfigOption<>(
                     "vertex.check_adjacent_vertex_exist",

--- a/hugegraph-core/src/main/java/com/baidu/hugegraph/schema/IndexLabel.java
+++ b/hugegraph-core/src/main/java/com/baidu/hugegraph/schema/IndexLabel.java
@@ -134,6 +134,26 @@ public class IndexLabel extends SchemaElement {
         return OLAP_ID.equals(this.baseValue);
     }
 
+    public Object validValue(Object value) {
+        if (!(value instanceof Number)) {
+            return value;
+        }
+
+        Number number = (Number) value;
+        switch (this.indexType()) {
+            case RANGE_INT:
+                return number.intValue();
+            case RANGE_LONG:
+                return number.longValue();
+            case RANGE_FLOAT:
+                return number.floatValue();
+            case RANGE_DOUBLE:
+                return number.doubleValue();
+            default:
+                return value;
+        }
+    }
+
     // ABS of System index id must be below SchemaElement.MAX_PRIMITIVE_SYS_ID
     private static final int VL_IL_ID = -1;
     private static final int EL_IL_ID = -2;

--- a/hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java
+++ b/hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java
@@ -519,8 +519,10 @@ public class VertexCoreTest extends BaseCoreTest {
 
         for (int i = 1; i <= testDataCount; i++) {
             graph.addVertex(T.label, "developer", "name", "developer" + i,
-                            "age", i * 3);
+                            "age", i * 3,
+                            "score", i * 3);
         }
+        graph.tx().commit();
 
         if (!removeLeftIndexOnOverwrite) {
             Whitebox.setInternalState(params().graphTransaction(),
@@ -530,6 +532,7 @@ public class VertexCoreTest extends BaseCoreTest {
         count = graph.traversal().V()
                      .hasLabel("developer")
                      .has("age", ConditionP.gt(0))
+                     .has("score", ConditionP.gt(0))
                      .limit(-1)
                      .count()
                      .next();

--- a/hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java
+++ b/hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java
@@ -509,6 +509,31 @@ public class VertexCoreTest extends BaseCoreTest {
                      .next();
         // Debug stop here to see whether the left index be correctly determined
         Assert.assertEquals(testDataCount, count);
+
+        // mock test removeLeftIndexIfNeeded
+        boolean removeLeftIndexOnOverwrite =
+                Whitebox.getInternalState(params().graphTransaction(),
+                                          "removeLeftIndexOnOverwrite");
+        Whitebox.setInternalState(params().graphTransaction(),
+                                  "removeLeftIndexOnOverwrite", true);
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "developer", "name", "developer" + i,
+                            "age", i * 3);
+        }
+
+        if (!removeLeftIndexOnOverwrite) {
+            Whitebox.setInternalState(params().graphTransaction(),
+                                      "removeLeftIndexOnOverwrite", false);
+        }
+
+        count = graph.traversal().V()
+                     .hasLabel("developer")
+                     .has("age", ConditionP.gt(0))
+                     .limit(-1)
+                     .count()
+                     .next();
+        Assert.assertEquals(testDataCount, count);
     }
 
     @Test

--- a/hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java
+++ b/hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java
@@ -414,6 +414,151 @@ public class VertexCoreTest extends BaseCoreTest {
     }
 
     @Test
+    public void testRemoveLeftRangeIndex() throws InterruptedException {
+        HugeGraph graph = graph();
+        SchemaManager schema = graph.schema();
+
+        schema.propertyKey("updateTime").asLong().create();
+
+        schema.vertexLabel("soft").properties("name", "updateTime")
+              .primaryKeys("name").create();
+
+        schema.indexLabel("softByUpdateTime").onV("soft").range()
+              .by("updateTime").create();
+
+        final long testDataCount = 10L;
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "soft", "name", "soft" + i,
+                            "updateTime", i);
+        }
+        graph.tx().commit();
+
+        long count = graph.traversal().V()
+                          .has("soft", "updateTime",
+                               ConditionP.gt(0))
+                          .limit(-1)
+                          .count()
+                          .next();
+        Assert.assertEquals(testDataCount, count);
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "soft", "name", "soft" + i,
+                            "updateTime", 2 * i);
+        }
+        graph.tx().commit();
+
+        count = graph.traversal().V()
+                     .has("soft", "updateTime",
+                          ConditionP.gt(0))
+                     .limit(-1)
+                     .count()
+                     .next();
+        Assert.assertEquals(testDataCount, count);
+
+        schema.propertyKey("score").asInt().create();
+        schema.vertexLabel("developer").properties("name", "age", "score")
+              .primaryKeys("name").create();
+
+        schema.indexLabel("developerByAge").onV("developer").range()
+              .by("age").create();
+        schema.indexLabel("developerByScore").onV("developer").range()
+              .by("score").create();
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "developer", "name", "developer" + i,
+                            "age", i,
+                            "score", i);
+        }
+        graph.tx().commit();
+
+        count = graph.traversal().V()
+                     .hasLabel("developer")
+                     .has("age", ConditionP.gt(0))
+                     .has("score", ConditionP.gt(0))
+                     .limit(-1)
+                     .count()
+                     .next();
+        Assert.assertEquals(testDataCount, count);
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "developer", "name", "developer" + i,
+                            "age", i * 2,
+                            "score", i * 2);
+        }
+        graph.tx().commit();
+
+        count = graph.traversal().V()
+                     .hasLabel("developer")
+                     .has("age", ConditionP.gt(0))
+                     .has("score", ConditionP.gt(0))
+                     .limit(-1)
+                     .count()
+                     .next();
+
+        Assert.assertEquals(testDataCount, count);
+
+        // Wait left index to be removed
+        Thread.sleep(2000);
+
+        count = graph.traversal().V()
+                     .hasLabel("developer")
+                     .has("age", ConditionP.gt(0))
+                     .limit(-1)
+                     .count()
+                     .next();
+        // Debug stop here to see whether the left index be correctly determined
+        Assert.assertEquals(testDataCount, count);
+    }
+
+    @Test
+    public void testLeftUnionIndex(){
+        HugeGraph graph = graph();
+        SchemaManager schema = graph.schema();
+
+        schema.propertyKey("updateTime").asLong().create();
+        schema.propertyKey("country").asText().create();
+
+        schema.vertexLabel("soft").properties("name", "country", "updateTime")
+              .primaryKeys("name").create();
+
+        schema.indexLabel("softByCountryAndUpdateTime").onV("soft").secondary()
+              .by("country", "updateTime").create();
+
+        final long testDataCount = 2L;
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "soft", "name", "soft" + i,
+                            "country", "china", "updateTime", i);
+        }
+        graph.tx().commit();
+
+        long count = graph.traversal().V()
+                          .hasLabel("soft")
+                          .has("updateTime", 2L)
+                          .has("country", "china")
+                          .limit(-1)
+                          .count()
+                          .next();
+        Assert.assertEquals(1, count);
+
+        for (int i = 1; i <= testDataCount; i++) {
+            graph.addVertex(T.label, "soft", "name", "soft" + i,
+                            "country", "china", "updateTime", i * 2);
+        }
+        graph.tx().commit();
+
+        count = graph.traversal().V()
+                     .hasLabel("soft")
+                     .has("updateTime", 2L)
+                     .has("country", "china")
+                     .limit(-1)
+                     .count()
+                     .next();
+        Assert.assertEquals(1L, count);
+    }
+
+    @Test
     public void testAddVertexWithPropertySet() {
         HugeGraph graph = graph();
         Vertex vertex = graph.addVertex(T.label, "review", "id", 1,

